third-culture-software/bhima
# Refactor inventory_unit to support translations for predefined types

Refactored inventory_unit to support translations for predefined types.  Added more useful predefined types and English and French translations.  This refactor also makes sure that the inventory_units are referred to as 'unit_type' everywhere in the front end (before it was 'unit' in some places and 'unit_type' in others).

Added a few more inventory_unit types than proposed in https://github.com/IMA-WorldHealth/bhima/issues/6741 to support common unit types used at our production sites.

Here is a list of the new Inventory Unit types:  [DEFAULT BHIMA Inventory Units 2022-08-16.pdf](https://github.com/IMA-WorldHealth/bhima/files/9352007/DEFAULT.BHIMA.Inventory.Units.2022-08-16.pdf)

OLD version with b/nnn inventory units: ~~[DEFAULT BHIMA Inventory Units 2022-08-15.pdf](https://github.com/IMA-WorldHealth/bhima/files/9349038/DEFAULT.BHIMA.Inventory.Units.2022-08-15.pdf)~~

Closes https://github.com/IMA-WorldHealth/bhima/issues/6741

**TESTING**
- Use bhima_test -- make sure to do:  yarn build:db
- Go to:  Inventory > Inventory Configuration
  - Look at the new values in the `Inventory Unit form` part (bottom right)
  - Note that predefined values cannot be edited or delete
  - Try adding a new value.  Note that it can be edited or deleted.
- Go to:  Inventory > Inventory Registry
  - Note the `Unit` column.  For bhima_test, these should all be 'unit' (or 'unité') -- except for code=MOT.HCRF250RX which should show the unit type as 'lot'.
  - Try changing the language to observing the effect
  - Edit one of the inventory items and select a different unit.  Save and then verify the change
- Check out any of the pages that use inventory units (eg, Shipments, Stock Exits, and documents for Stock Exits) to verify that inventory units are handled correctlu

**MIGRATION OF PRODUCTION SITES!**
I have verified the migration behavior on one production site (Vanga, which is at the previous version of Bhima).  There are site-specific migration scripts for all the other active production sites.   All but Vanga are untested, but should work.  **WARNING** 
  - It is **VERY IMPORTANT** to make a backup of the site's database before trying to upgrade.  There is a significant chance that the migration scripts may need to be updated before they can be applied successfully.
  - You must determine which version of BHIMA is running on the production site you are upgrading
  - Do 'yarn migrate' to generate the full migration script.  Recall that 'yarn migrate' creates a full migration script -- including the site-specific part in separate files for each site.
  - After doing 'yarn migrate', the resulting full migration script will need to edited to insert the migration scripts between the release version in effect for the site and the current release in the appropriate place.   
  - ALSO, you may need to move the shipment specific triggers (shipment_reference, shipment_document_map) to a point in the full migration script AFTER the shipment table has been created!   BEFORE applying the script, check the full migration script to see if this change is necessary.
